### PR TITLE
feat(plugin-wasm): add async-wasm support

### PR DESCRIPTION
**Pull request by tuchg**
## Summary

Async wasm support has been added, along with partial support for loading wasm using "fetch". 

However, sync WebAssembly support has not yet been added. 

The `wasm-parser` crate is used to parse WebAssembly ast, and the currently supported webpack-related tests have been successfully ported.

## Related issue (if exists)

* #2158
* async-modules support #[d5b54e5](https://github.com/web-infra-dev/rspack/pull/2229/commits/d5b54e5e759a09456a2324c98ef56458a3a07406)
* #2463

## Types of changes

- [ ] Docs change / Dependency upgrade
- [ ] Bug fix
- [x] New feature / Improvement
- [ ] Refactoring
- [ ] Breaking change

## Checklist

- [ ] I have added changeset via `pnpm run changeset`.
- [x] I have added tests to cover my changes.

### CODE REVIEW

**Comment by hardfist** [comment]
I'm shocked!

**Comment by zoolsher** [comment]
You rock!

**Comment by ahabhgk** [comment]
Feel free to ping me once this is ready to review @tuchg

**Comment by tuchg** [comment]
@ahabhgk hi, it's done.

File: crates/rspack_plugin_javascript/src/plugin.rs [hunk 1]
```diff
@@ -573,3 +576,60 @@ impl Plugin for JsPlugin {
...
+#[derive(Debug)]
+pub struct InferAsyncModulesPlugin;
+
+#[async_trait::async_trait]
+impl Plugin for InferAsyncModulesPlugin {
+  fn name(&self) -> &'static str {
+    "InferAsyncModulesPlugin"
+  }
+
+  fn module_ids(&mut self, compilation: &mut Compilation) -> Result<()> {
```
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/plugin.rs:589]: This hook should be `finishModules`, you can add `finishModules` on `Plugin` trait in this PR or a following PR.

File: crates/rspack_plugin_javascript/src/visitors/async_module.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+
+    let last_import = items
+      .iter()
+      .enumerate()
+      .skip_while(|(_, item)| !matches!(item, ModuleItem::Stmt(Stmt::Decl(Decl::Var(_)))))
+      .take_while(|(_, item)| matches!(item, ModuleItem::Stmt(Stmt::Decl(Decl::Var(_)))))
+      .map(|(i, item)| {
+        if let Some(is_async) = self.promises.pop_front() && is_async {
+          if let ModuleItem::Stmt(Stmt::Decl(Decl::Var(var))) = item {
+            let decl = unsafe { var.decls.get_unchecked(0) };
```
Comment (ahabhgk) [inline crates/rspack_plugin_javascript/src/visitors/async_module.rs:39]: Need to add safety comment

**Review by ahabhgk — APPROVED** [review]
Thank you!

**Pull request merged**